File: cpp/22_generate_parentheses.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  void generateParen(vector<string> &result, string paren, int open,
                     int close) {
    if (open == 0 && close == 0) {
      result.push_back(std::move(paren));
      return;
    }
    if (open > 0) {
      generateParen(result, paren + "(", open - 1, close);
    }
    if (close > open) {
      generateParen(result, paren + ")", open, close - 1);
    }
  }
  vector<string> generateParenthesis(int n) {
    vector<string> result;
    generateParen(result, "", n, n);
    return result;
  }
};

File: rust/22_generate_parentheses.rs
impl Solution {
    fn generate_helper(result: &mut Vec<String>, paren: &mut String, open: i32, close: i32) {
        if open == 0 && close == 0 {
            result.push(paren.clone());
            return;
        }

        if open > 0 {
            paren.push('(');
            Solution::generate_helper(result, paren, open - 1, close);
            paren.pop();
        }
        if close > open {
            paren.push(')');
            Solution::generate_helper(result, paren, open, close - 1);
            paren.pop();
        }
    }

    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let mut result = vec![];
        let mut paren = "".to_string();
        Solution::generate_helper(&mut result, &mut paren, n, n);
        result
    }
}
